codec-http2/src/main/java/io/netty/handler/codec/http2/ReadOnlyHttp2Headers.java

```diff
@@ -0,0 +1,843 @@
+                                                     AsciiString authority, AsciiString[] otherHeaders,
+                                                     boolean validateHeaders) {
+        return new ReadOnlyHttp2Headers(new AsciiString[] {
+                PseudoHeaderName.METHOD.value(), method, PseudoHeaderName.PATH.value(), path,
+                PseudoHeaderName.SCHEME.value(), scheme, PseudoHeaderName.AUTHORITY.value(), authority
+                }, otherHeaders, validateHeaders);
+    }
+
+    /**
+     * Create a new read only representation of headers used by servers.
+     * @param status The value for {@link PseudoHeaderName#STATUS}.
+     * @param otherHeaders A an array of key:value pairs. Must not contain any
+     *                     <a href="https://tools.ietf.org/html/rfc7540#section-8.1.2.1">pseudo headers</a>
+     *                     or {@code null} names/values.
+     *                     A copy will <strong>NOT</strong> be made of this array. If the contents of this array
+     *                     may be modified externally you are responsible for passing in a copy.
+     * @param validateHeaders {@code true} will run validation on each header name/value pair to ensure protocol
+     *                        compliance.
+     * @return a new read only representation of headers used by servers.
+     */
+    public static ReadOnlyHttp2Headers serverHeaders(AsciiString status, AsciiString[] otherHeaders,
+                                                     boolean validateHeaders) {
+        return new ReadOnlyHttp2Headers(new AsciiString[] { PseudoHeaderName.STATUS.value(), status }, otherHeaders,
+                                        validateHeaders);
+    }
+
+    private ReadOnlyHttp2Headers(AsciiString[] pseudoHeaders, AsciiString[] otherHeaders, boolean validateHeaders) {
+        if ((pseudoHeaders.length & 1) != 0 || (otherHeaders.length & 1) != 0) {
+            throw newIllegalArgumentException();
+        }
+        if (validateHeaders) {
+            validateHeaders(pseudoHeaders, otherHeaders);
+        }
+        this.pseudoHeaders = pseudoHeaders;
+        this.otherHeaders = otherHeaders;
+    }
+
+    private static IllegalArgumentException newIllegalArgumentException() {
+        return new IllegalArgumentException("pseudoHeaders and otherHeaders must be arrays of [name, value] pairs");
+    }
+
+    private static void validateHeaders(AsciiString[] pseudoHeaders, AsciiString[] otherHeaders) {
+        boolean seenNonPseudoHeader = false;
+        int i;
+        AsciiString name;
+        for (i = 0; i < pseudoHeaders.length; ++i) {
+            name = pseudoHeaders[i];
+            HTTP2_NAME_VALIDATOR.validateName(name);
+            if (!seenNonPseudoHeader && !name.isEmpty() && name.byteAt(0) != ':') {
+                seenNonPseudoHeader = true;
+            } else if (seenNonPseudoHeader && !name.isEmpty() && name.byteAt(0) == ':') {
```

Comment:
This seems like doing extra work.  In the case that the caller used one of the special constructors, the pseudo headers are always going to be valid right?  It would be a waste to check them here.